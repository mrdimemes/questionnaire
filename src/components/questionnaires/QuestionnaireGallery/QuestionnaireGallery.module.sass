@use "src/sass/foundation/breakpoints"

.QuestionnaireGallery
  gap: 10px
  padding-bottom: 100px
  position: relative

.viewSwitch
  position: fixed
  top: 90px
  right: 60px
  @media (max-width: breakpoints.$small-screen)
    top: 80px
    right: 50px
  @media (max-width: breakpoints.$tablet)
    top: 60px
    right: 30px
  @media (max-width: breakpoints.$mobile)
    right: 20px

.pagination
  position: absolute
  bottom: 0
  left: 0
  right: 0