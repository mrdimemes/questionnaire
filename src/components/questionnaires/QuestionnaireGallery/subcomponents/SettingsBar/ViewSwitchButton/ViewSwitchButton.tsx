import classNames from "classnames";

import { Button } from "src/components";
import { withTheme } from "src/HOCs";

import { GalleryView } from "../../../models";

import styles from "./ViewSwitchButton.module.sass";

import type { ViewSwitchProps } from "./types";


const ViewSwitchButton = ({
  setCurrentView,
  currentView,
  className,
}: ViewSwitchProps) => {

  const toggleView = () => {
    setCurrentView(
      currentView === GalleryView.Plates ?
        GalleryView.Rows :
        GalleryView.Plates,
    );
  };

  return (
    <Button
      onClick={toggleView}
      className={classNames(
        styles.ViewSwitchButton,
        currentView === GalleryView.Plates ? styles.plates : styles.rows,
        className,
      )}
    />
  );
};

export default withTheme(ViewSwitchButton, styles);