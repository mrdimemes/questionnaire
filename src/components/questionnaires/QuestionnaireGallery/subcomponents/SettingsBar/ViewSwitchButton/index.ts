export { default as ViewSwitchButton } from "./ViewSwitchButton";