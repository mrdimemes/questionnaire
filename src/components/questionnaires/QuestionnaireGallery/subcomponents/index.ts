export { QuestionnaireCardComponent } from "./QuestionnaireCardComponent";
export { ViewSwitchButton } from "./ViewSwitchButton";
export { PlatesView } from "./PlatesView";
export { RowsView } from "./RowsView";