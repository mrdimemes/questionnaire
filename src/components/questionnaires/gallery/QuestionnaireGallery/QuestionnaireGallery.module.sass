.body
  gap: 10px

.view_rows
  display: flex
  flex-direction: column

.view_plates
  display: grid
  grid-template-columns: 1fr 1fr

.column
  display: flex
  flex-direction: column
  gap: 10px