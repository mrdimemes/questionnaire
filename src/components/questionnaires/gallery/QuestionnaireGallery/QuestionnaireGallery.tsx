import classNames from "classnames";
import { useState, useEffect } from "react";
import axios from "axios";
import { FetchStatus } from "src/models/FetchStatus";
import { QuestionnaireCard } from "src/models/questionnaire/QuestionnaireCard";
import { QuestionnaireCardComponent, ViewSwitchButton } from "../";
import { LoadingSpinner } from "src/components/LoadingSpinner";
import styles from "./QuestionnaireGallery.module.sass";

export enum GalleryViews {
  Rows = "ROWS",
  Plates = "PLATES"
}

const QuestionnaireGallery = () => {
  const [status, setStatus] = useState(FetchStatus.Loading);
  const [cards, setCards] = useState<Array<QuestionnaireCard>>();
  const [currentView, setCurrentView] = useState(GalleryViews.Rows);

  const getCardNode = (card: QuestionnaireCard) => {
    return <QuestionnaireCardComponent
      key={card.id}
      id={card.id}
      label={card.label}
      tags={card.tags}
    />
  };

  useEffect(() => {
    axios.get(process.env.REACT_APP_BACKEND_URL +
      "questionnaires/questionnaireCards").then(
        (res) => {
          setCards(res.data.questionnaires);
          setStatus(FetchStatus.Complete);
        }
      )
  }, []);

  return (
    <div className={classNames(
      styles.body,
      currentView === GalleryViews.Rows ? styles.view_rows : styles.view_plates
    )}>

      {status === FetchStatus.Loading && <LoadingSpinner />}

      {status === FetchStatus.Complete &&
        <>
          <ViewSwitchButton
            setCurrentView={setCurrentView} currentView={currentView}
          />

          {currentView === GalleryViews.Plates && <>
            <div className={styles.column}> {
              cards?.filter((_, index) => index % 2 === 0)
                .map((card) => getCardNode(card))
            } </div>
            <div className={styles.column}> {
              cards?.filter((_, index) => index % 2 !== 0)
                .map((card) => getCardNode(card))
            } </div>
          </>}

          {currentView === GalleryViews.Rows &&
            cards?.map((card) => getCardNode(card))}
        </>
      }

    </div >
  )
}

export default QuestionnaireGallery