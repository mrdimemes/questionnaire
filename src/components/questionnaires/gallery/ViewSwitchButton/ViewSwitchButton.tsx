import classNames from "classnames";
import { useAppSelector } from "src/redux/hooks";
import { themeSelector } from "src/redux/selectors";
import { getThemeStyle } from "src/redux/slices/themeSlice";
import { Button } from "src/components/UI/Button";
import { GalleryViews } from "../QuestionnaireGallery/QuestionnaireGallery";
import Plates from "src/assets/images/icons/plates.svg";
import Rows from "src/assets/images/icons/rows.svg";
import styles from "./ViewSwitchButton.module.sass";

type ViewSwitchProps = {
  setCurrentView: (newView: GalleryViews) => void,
  currentView: GalleryViews
};

const ViewSwitchButton = (
  { setCurrentView, currentView }: ViewSwitchProps
) => {
  const currentTheme = useAppSelector(themeSelector);
  const toggleView = () => {
    switch (currentView) {
      case GalleryViews.Plates:
        setCurrentView(GalleryViews.Rows);
        break;
      case GalleryViews.Rows:
        setCurrentView(GalleryViews.Plates)
        break;
    }
  };

  return (
    <Button
      onClick={toggleView}
      className={classNames(
        styles.body,
        getThemeStyle(styles, currentTheme)
      )}
    >
      <img
        src={currentView === GalleryViews.Plates ? Rows : Plates}
        alt="view mode"
      />
    </Button>
  )
}

export default ViewSwitchButton