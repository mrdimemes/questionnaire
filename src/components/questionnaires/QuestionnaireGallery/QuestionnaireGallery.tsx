import { useState, useCallback, useEffect } from "react";

import { QuestionnaireCard, SortOption } from "src/models";
import { QuestionnaireService } from "src/services";
import { PaginationBar, Loadable } from "src/components";

import { PlatesView, RowsView, SettingsBar } from "./subcomponents";
import { GalleryView } from "./models";
import styles from "./QuestionnaireGallery.module.sass";

import type { QuestionnaireGalleryProps } from "./types";


const QuestionnaireGallery = ({ filterTag }: QuestionnaireGalleryProps) => {
  const [cards, setCards] = useState<QuestionnaireCard[]>([]);
  const [currentView, setCurrentView] = useState(GalleryView.Rows);
  const [sortOption, setSortOption] = useState(SortOption.NoSort);
  const [searchPhrase, setSearchPhrase] = useState("");
  const [activePage, setActivePage] = useState(1);
  const [totalPages, setTotalPages] = useState(1);
  const cardsPerPage = 10;

  const loadPage = useCallback(
    async () => {
      const cardBunch = await QuestionnaireService
        .getQuestionnaireCards(
          sortOption,
          searchPhrase,
          filterTag?.id ?? null,
          activePage,
          cardsPerPage,
        );
      setCards(cardBunch.cards);
      setTotalPages(cardBunch.totalPages);
    },
    [activePage, filterTag, searchPhrase, sortOption],
  );

  useEffect(() => setActivePage(1), [sortOption, filterTag, searchPhrase]);

  return (
    <Loadable load={loadPage}>
      <div className={styles.QuestionnaireGallery}>

        <SettingsBar
          sortOption={sortOption}
          setSortOption={setSortOption}
          currentView={currentView}
          setCurrentView={setCurrentView}
          searchPhrase={searchPhrase}
          setSearchPhrase={setSearchPhrase}
        />

        {
          filterTag &&
          <p className={styles.prompt}>
            Опросы, содержащие тег «{filterTag.label}».
          </p>
        }

        {
          cards.length === 0 &&
          <p className={styles.prompt}>По запросу ничего не найдено.</p>
        }

        {currentView === GalleryView.Plates && <PlatesView cards={cards} />}
        {currentView === GalleryView.Rows && <RowsView cards={cards} />}

        <PaginationBar
          className={styles.pagination}
          activePage={activePage}
          totalPages={totalPages ?? 1}
          callback={setActivePage}
        />

      </div >
    </Loadable>
  );
};

export default QuestionnaireGallery;