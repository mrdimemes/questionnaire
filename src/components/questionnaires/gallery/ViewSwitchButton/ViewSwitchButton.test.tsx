import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithProviders } from "src/utils/test-utils";
import { GalleryViews } from "../QuestionnaireGallery/QuestionnaireGallery";
import ViewSwitchButton from "./ViewSwitchButton";

let currentView: GalleryViews;
const setCurrentView = (newView: GalleryViews) => {
  currentView = newView;
}

beforeEach(() => setCurrentView(GalleryViews.Rows));

describe("QuestionnaireGallery/ViewSwitchButton component", () => {
  test("ViewSwitchButton renders", () => {
    renderWithProviders(
      <ViewSwitchButton
        setCurrentView={setCurrentView}
        currentView={currentView}
      />
    );
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  test("ViewSwitchButton switches the view", () => {
    renderWithProviders(
      <ViewSwitchButton
        setCurrentView={setCurrentView}
        currentView={currentView}
      />
    );
    userEvent.click(screen.getByRole("button"));
    expect(currentView).toEqual(GalleryViews.Plates);
  });

  test("ViewSwitchButton snapshot", () => {
    const { container } = renderWithProviders(
      <ViewSwitchButton
        setCurrentView={setCurrentView}
        currentView={currentView}
      />
    );
    expect(container).toMatchSnapshot();
  });
});