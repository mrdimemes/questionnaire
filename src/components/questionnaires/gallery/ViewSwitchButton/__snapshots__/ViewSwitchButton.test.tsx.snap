// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionnaireGallery/ViewSwitchButton component ViewSwitchButton snapshot 1`] = `
<div>
  <button
    class="body theme_light body theme_light"
  >
    <img
      alt="view mode"
      src="plates.svg"
    />
  </button>
</div>
`;
